### PR TITLE
Added scriptstate JSON formatting

### PR DESCRIPTION
**Pull request by Chr1Z93**
The LuaScriptState field contains a string that contains information saved by a Luascript. This string will in most cases be the result of JSON.encode().

This PR adds JSON formatting to the created .luascriptstate file IF it is an actual JSON string.

### CODE REVIEW

File: mod/reverse.go [hunk 1]
```diff
@@ -18,9 +19,7 @@ type Reverser struct {
 	ObjWriter         file.JSONWriter
 	ObjDirCreator     file.DirCreator
 	RootWrite         file.JSONWriter
-
-	// If not empty: holds the entire filename (C:...) of the json to read
-	OnlyObjState      string
+	OnlyObjState      string // If not empty: holds the entire filename (C:...) of the json to read
```
Comment (argonui) [inline mod/reverse.go:22]: Go style has comments on preceding line

File: main.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func main() {
 	// handling for saved objects instead of a full savegame
 	if *objin != "" {
 		objs = file.NewJSONOps(filepath.Dir(*objin))
-		objdir = file.NewDirOps(filepath.Dir(*objin))
+		objdir = file.NewDirOps(filepath.Dir(*objout))
```
Comment (argonui) [inline main.go:60]: Let's break this into it's own change. I'm not sure that it is a bug. And if it is, then that means the whole feature is not well tested

File: file/jsonops.go [hunk 3]
```diff
@@ -62,7 +63,22 @@ func (j *JSONOps) pullRawFile(filename string) ([]byte, error) {
 	}
 	defer jFile.Close()
 
-	return ioutil.ReadAll(jFile)
+	return io.ReadAll(jFile)
+}
+
+// WriteJSON writes a serialized JSON object or array to a file.
+func (j *JSONOps) WriteJSON(m interface{}, filename string) error {
```
Comment (argonui) [inline file/jsonops.go:70]: This is a duplicate of write obj. Use the existing function

File: file/jsonops.go [hunk 4]
```diff
@@ -103,7 +119,7 @@ func ReadRawFile(filename string) (map[string]interface{}, error) {
 	}
 	defer jFile.Close()
 
-	b, err := ioutil.ReadAll(jFile)
```
Comment (argonui) [inline file/jsonops.go:106]: Break this into it's own pr. Changing the underlying io operations is something that we should be ready to rollback

File: mod/reverse.go [hunk 5]
```diff
@@ -67,11 +66,23 @@ func (r *Reverser) Write(raw map[string]interface{}) error {
...
-
-		err := r.LuaWriter.EncodeToFile(strVal, createdFile)
-		if err != nil {
-			return fmt.Errorf("lua.EncodeToFile(<value>, %s) : %v", createdFile, err)
+		
+		var jsonInterface map[string]interface{}
+		err := json.Unmarshal([]byte(strVal), &jsonInterface)
+		if err == nil {
+			// if it's JSON, use the JSONWriter
+			err := r.ObjWriter.WriteJSON(jsonInterface, createdFile)
```
Comment (argonui) [inline mod/reverse.go:74]: This does more than format. You're writing a new file? How often is the script state illegible as a string?

**Review by argonui — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (Chr1Z93) [reply main.go:60]: Without this change, there will be empty folders created in the folder of the `objin` file that mimic the actual folder structure of the output.

File: mod/reverse.go (hunk 1 above)
Comment (Chr1Z93) [reply mod/reverse.go:22]: Thanks, corrected it

File: file/jsonops.go (hunk 4 above)
Comment (Chr1Z93) [reply file/jsonops.go:106]: That's fair but afaik there shouldn't be a need for that since with Go 1.16 ioutil.ReadAll was changed to io.ReadAll (https://github.com/golang/go/issues/40025)

File: mod/reverse.go (hunk 5 above)
Comment (Chr1Z93) [reply mod/reverse.go:76]: Script state as string is pretty much always illegible and it also sucks for source control since it's a single line. See here for a recent example: https://github.com/argonui/SCED/pull/786/files#diff-34b51a0c82f85d6d9927ddc988df3589947b8c241d04199af6a4dfbc3c3f6044

File: file/jsonops.go (hunk 3 above)
Comment (Chr1Z93) [reply file/jsonops.go]: Updated PR

File: go.mod [hunk 6]
```diff
@@ -1,6 +1,6 @@
 module ModCreator
 
-go 1.17
+go 1.18
```
Comment (Chr1Z93) [inline go.mod:3]: Required for using "any" in types.J

File: compare_test.go [hunk 7]
```diff
@@ -83,7 +83,7 @@ func toObjArray(i interface{}) ([]map[string]interface{}, error) {
 	return arr, nil
 }
 
-func compareObjArrays(t *testing.T, guid string, a, b []map[string]interface{}) error {
+func compareObjArrays(t *testing.T, a, b []map[string]interface{}) error {
```
Comment (Chr1Z93) [inline compare_test.go:86]: `guid` was unused in this function and thus obsolete
